app/src/main/java/com/example/grocerieswizard/ui/home/HomeFragment.java

```diff
@@ -249,4 +174,73 @@ private String getStringIngredients(List<IngredientUi> ingredients) {
         return stringBuilder.toString();
     }
 
+    @Override
+    public void showRecipes(List<RecipeUi> recipes) {
+        adapter.setRecipeList(recipes);
+    }
+
+    private void showRecipeDetails(RecipeUi recipe) {
+        if (recipe != null) {
+            if (!recipe.isSwiped()) {
+                recipe.setSwiped(false);
+                adapter.itemChanged(adapter.getPositionForRecipe(recipe));
+                DetailFragment detailFragment = DetailFragment.newInstance(recipe);
+                requireActivity().getSupportFragmentManager()
+                        .beginTransaction()
+                        .replace(R.id.frameLayout, detailFragment)
+                        .addToBackStack(null)
+                        .commit();
+            } else {
+                Log.d("MainActivity", "recipe model null");
+            }
+        }
+    }
+
+    @Override
+    public void showDeleteConfirmation(RecipeUi recipe) {
+            AlertDialog.Builder builder = new AlertDialog.Builder(requireContext());
+            builder.setTitle("Confirm Deletion");
+            builder.setMessage("Are you sure you want to delete " + recipe.getRecipeName() + " recipe?");
+            builder.setPositiveButton("Delete", (dialog, which) -> {
+                // User confirmed deletion, remove the recipe and update the RecyclerView
+                adapter.removeRecipe(recipe);
+                Toast.makeText(requireContext(), "Recipe deleted: " + recipe.getRecipeName(), Toast.LENGTH_SHORT).show();
+                dialog.dismiss();
+            });
+            builder.setNegativeButton("Cancel", (dialog, which) -> {
+                recipe.setSwiped(false);
+                adapter.itemChanged(adapter.getPositionForRecipe(recipe));
+                dialog.dismiss();
+            });
+            // Create and show the dialog
+            AlertDialog alertDialog = builder.create();
+            alertDialog.show();
+    }
+
+    @Override
+    public void showEditRecipe(RecipeUi recipe) {
+        FragmentTransaction transaction = requireActivity().getSupportFragmentManager().beginTransaction();
+        AddRecipeFragment addRecipeFragment = new AddRecipeFragment();
```

Comment:
When creating fragment with some arguments, create static function named "newInstance" in the fragment so here you should be able to create the fragment like this:

```
AddRecipeFragment fragment = AddRecipeFragment.newInstance(recipe, position);
```